eureka-client/src/main/java/com/netflix/discovery/shared/JerseyEurekaHttpClient.java

```diff
@@ -0,0 +1,153 @@
+
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response.Status;
+
+import com.netflix.appinfo.InstanceInfo;
+import com.netflix.appinfo.InstanceInfo.InstanceStatus;
+import com.sun.jersey.api.client.ClientResponse;
+import com.sun.jersey.api.client.WebResource;
+import com.sun.jersey.client.apache4.ApacheHttpClient4;
+
+/**
+ * @author Tomasz Bak
+ */
+public abstract class JerseyEurekaHttpClient implements EurekaHttpClient {
+
+    protected final String serviceUrl;
+
+    protected JerseyEurekaHttpClient(String serviceUrl) {
+        this.serviceUrl = serviceUrl;
+    }
+
+    protected abstract ApacheHttpClient4 getJerseyApacheClient();
+
+    @Override
+    public HttpResponse<Void> register(InstanceInfo info) {
+        String urlPath = "apps/" + info.getAppName();
+        ClientResponse response = null;
+        try {
+            WebResource webResource = getJerseyApacheClient().resource(serviceUrl).path(urlPath);
+            addExtraHeaders(webResource);
+            response = webResource
+                    .header("Accept-Encoding", "gzip")
+                    .type(MediaType.APPLICATION_JSON_TYPE)
+                    .accept(MediaType.APPLICATION_JSON)
+                    .post(ClientResponse.class, info);
+            return HttpResponse.responseWith(response.getStatus());
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+        }
+    }
+
+    @Override
+    public HttpResponse<Void> cancel(String appName, String id) {
+        ClientResponse response = null;
+        try {
+            String urlPath = "apps/" + appName + "/" + id;
+            WebResource webResource = getJerseyApacheClient().resource(serviceUrl).path(urlPath);
+            addExtraHeaders(webResource);
+            response = webResource.delete(ClientResponse.class);
+            return HttpResponse.responseWith(response.getStatus());
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+        }
+    }
+
+    @Override
+    public HttpResponse<InstanceInfo> sendHeartBeat(String appName, String id, InstanceInfo info, InstanceStatus overriddenStatus) {
+        ClientResponse response = null;
+        try {
+            String urlPath = "apps/" + appName + "/" + id;
+            WebResource webResource = getJerseyApacheClient().resource(serviceUrl)
+                    .path(urlPath)
+                    .queryParam("status", info.getStatus().toString())
+                    .queryParam("lastDirtyTimestamp", info.getLastDirtyTimestamp().toString());
+            if (overriddenStatus != null) {
+                webResource = webResource.queryParam("overriddenstatus", overriddenStatus.name());
+            }
+            addExtraHeaders(webResource);
+            response = webResource.accept(MediaType.APPLICATION_JSON_TYPE).put(ClientResponse.class);
+            InstanceInfo infoFromPeer = null;
```

Comment:
Good point. We need to override this implementation in replicating client.

On Tue, Jun 2, 2015 at 5:42 PM, qiangdavidliu <notifications@github.com>
wrote:

> In
> eureka-client/src/main/java/com/netflix/discovery/shared/JerseyEurekaHttpClient.java
> <https://github.com/Netflix/eureka/pull/529#discussion_r31585722>:
>
> > +
> > +    @Override
> > +    public HttpResponse<InstanceInfo> sendHeartBeat(String appName, String id, InstanceInfo info, InstanceStatus overriddenStatus) {
> > +        ClientResponse response = null;
> > +        try {
> > +            String urlPath = "apps/" + appName + "/" + id;
> > +            WebResource webResource = getJerseyApacheClient().resource(serviceUrl)
> > +                    .path(urlPath)
> > +                    .queryParam("status", info.getStatus().toString())
> > +                    .queryParam("lastDirtyTimestamp", info.getLastDirtyTimestamp().toString());
> > +            if (overriddenStatus != null) {
> > +                webResource = webResource.queryParam("overriddenstatus", overriddenStatus.name());
> > +            }
> > +            addExtraHeaders(webResource);
> > +            response = webResource.accept(MediaType.APPLICATION_JSON_TYPE).put(ClientResponse.class);
> > +            InstanceInfo infoFromPeer = null;
>
> Can we move these peer specific code to a server implementation of this
> class? If we can break these out, it would be nice to refactor
> DiscoveryClient with this class.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/eureka/pull/529/files#r31585722>.
>